Transpile C API

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds the initial transpiler C API to Qiskit, it provides a
function for running the transpiler to compile a single circuit for a
provided target. This function is not designed to replace the Python
transpiler and should only be used with circuits defined solely using
the C API. It is backed by the same Rust implementation of the
transpiler passes, but when coming from the C/Rust API several
assumptions are made that the circuits only contain the feature set
defined in the C API and the results are only valid when called
with circuits built solely using the C API.

As this function is the initial version of the transpiler in C/Rust
there are still a few limitations beyond just the circuits come from C
and only have the features exposed to C. The first is that multi qubit
unitary synthesis does not exist in C yet, so if you construct any
3 or more qubit UnitaryGates the transpiler will error. Additionally,
there is no scheduling stage yet in the C transpiler yet, so there
are no options for scheduling circuits available. Finally, there is no
pluggability to the transpiler, unlike Python where the preset pass
managers and the transpile() function provide a great deal of
configurability and customization via composing your own passes or
using the plugin interface, these concepts do not exist in the C
transpiler. The limitations will be addressed in future releases,
but this first function is about enabling initial circuit
compilation support from C.


### Details and comments

Closes #14245

This is very much blocked until most of he sub-issues on #14245 are implemented. This does *not* compile and won't be able to for sometime. 

The current list of dependencies for this PR are:

- [x] #14835 
- [x] #14836 
- [x] #14768
- [x] #14711
  * [x] Depends on #14780 
- [x] #14904 
  * [x] Depends on #14826
- [x] #14921 
- [x] #14885 
- [x] #14847 
- [x] #14780 
- [x] #14720
- [x] #14659

While waiting for these PRs you can look at #14922 which speculatively merged all these in progress PRs ontop of this branch and is showing this PR working.

mtreinish committed:
Add transpiler C API

This commit adds the initial transpiler C API to Qiskit, it provides a
function for running the transpiler to compile a single circuit for a
provided target. This function is not designed to replace the Python
transpiler and should only be used with circuits defined solely using
the C API. It is backed by the same Rust implementation of the
trnaspiler passes, but when coming from the C/Rust API several
assumptions are made that the circuits only contain the feature set
defined in the C API and the results are only valid when called
with circuits built solely using the C API.

As this function is the initial version of the transpiler in C/Rust
there are still a few limitations beyond just the circuits come from C
and only have the features exposed to C. The first is that multi qubit
unitary synthesis does not exist in C yet, so if you construct any
3 or more qubit UnitaryGates the transpiler will error. Additionally,
there is no scheduling stage yet in the C transpiler yet, so there
are no options for scheduling circuits available. Finally, there is no
pluggability to the transpiler, unlike Python where the preset pass
managers and the transpile() function provide a great deal of
configurability and customization via composing your own passes or
using the plugin interface, these concepts do no exist in the C
transpiler. The limitations will be addressed in future releases,
but this first function is about enabling initial circuit
compilation support from C.

mtreinish commented on crates/cext/src/transpiler/transpiler.rs:
```diff
@@ -0,0 +1,45 @@
+    target: *const Target,
+    optimization_level: u8,
+    seed: i64,
+) -> TranspileResult {
```
```suggestion
pub unsafe extern "C" fn qk_transpile(
    qc: *const CiruitData,
    target: *const Target,
    options: *const TranspileOptions,
    result: *mut TranspileResult,
    error: *mut *mut c_char,
) -> i32 {
```

mtreinish commented:
With the last blocker PR merged this is now ready for review. I will work on a follow up PR that does more correctness testing of from Python by using ctypes to build and compile circuits using the C API and then converting those to Python circuits to use `Operator.from_circuit` and other tools to do deeper validation. The testing here is just some small circuits and very high level assertions that the circuits are valid ISA circuits.

The only thing missing from the preset pass managers here that is in Python is `VF2PostLayout` which isn't available in Rust yet because it's at the top of a long PR stack that needs to be reviewed. If that's finished for 2.2 we can add it to this PR, if not we can add it in 2.3.

mtreinish commented:
~Hmm, there is a bug in optimization level 0 that I need to get to the bottom of. But it looks like level 1 - 3 are working.~

EDIT: fixed by https://github.com/Qiskit/qiskit/pull/14760/commits/d7ca189cd2636a48af918c073f0dce456994209a

mtreinish committed:
Add docs files

mtreinish committed:
Fix routing and level 0

mtreinish committed:
Fix clippy

mtreinish commented:
Hmm, I bet the segfault on windows is the same failure that #14751 is hitting that #14873 is fixing. If it persists after #14873 merges I'll dig more into it.

mtreinish committed:
Merge branch 'main' into transpile-c-wip

mtreinish committed:
Fix VF2 layout allocation with idle qubits

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>
Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

jakelishman commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+/// for execution.
+///
+/// This function is multithreaded internally and will launch a thread pool
+/// with threads equal to the number of CPUs by default. You can tune the
```
For HPC contexts, and leaving aside the standard stuff of "we need to offer a better thread-control API", I'm not certain what "equal to the number of CPUs by default" means precisely. Let's clarify this to be "the number of (physical / hyperthreaded) cores available (on the system as a whole / available to the process)" or whatever's appropriate.

mtreinish replied:
It's actually kernel reported CPUs, the exact behavior of what those CPUs are depends on your OS and your hardware. The rayon docs just say the number of CPUs available and do make a note about "hyperthreading": https://github.com/rayon-rs/rayon/blob/main/FAQ.md#how-many-threads-will-rayon-spawn . I can tweak the wording here to be more precise.

mtreinish replied:
Reworded in: https://github.com/Qiskit/qiskit/pull/14760/commits/e5cbe55511ff1f669401b482411b22d33706df5f

jakelishman commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+pub struct TranspileResult {
+    circuit: *mut CircuitData,
+    layout: *mut TranspileLayout,
+}
```
Since we're expecting users to allocate space for this themselves, it would be good to comment about a standard initialisation pattern for it, including that the internal pointers can be null or otherwise unset.

mtreinish replied:
I expanded on this in: https://github.com/Qiskit/qiskit/pull/14760/commits/e5cbe55511ff1f669401b482411b22d33706df5f

jakelishman commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+///   to this pointer. That pointer needs to be freed with `qk_str_free`.
+///
+/// @returns the return code for the transpiler, 0 means success and all other values indicate an
+///   error
```
With `result`, I think we should talk more about the ownership of the resulting objects, and who's responsible for freeing them and under what conditions.  It'd probably also be good to point out that we typically expect somebody to "move" ownership of the `CircuitData` and `TranspileLayout` out of `TranspileResult` and use them separately, rather than pretending that they're "owned" by that struct (though tbh, I don't know if that'd already be obvious to C programmers).

mtreinish replied:
I expanded on this a bit in: https://github.com/Qiskit/qiskit/pull/14760/commits/e5cbe55511ff1f669401b482411b22d33706df5f

jakelishman commented on crates/transpiler/src/passes/sabre/mod.rs:
```diff
@@ -16,12 +16,15 @@ pub mod heuristic;
 mod layout;
 mod neighbors;
-mod route;
+pub(crate) mod route;
 
 use pyo3::prelude::*;
 use pyo3::wrap_pyfunction;
 
+pub(crate) use heuristic::Heuristic;
+pub(crate) use heuristic::SetScaling;
 pub use layout::sabre_layout_and_routing;
+pub(crate) use route::sabre_routing;
```
Largely unimportant, but `route` probably needn't be `pub` at all, and anything that's `pub(crate) use` can/ought to be full `pub use`.

*edit*: we can follow up on this - we're going to need to rework the Sabre interfaces to use `TranspileLayout` anyway.

jakelishman commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -316,7 +316,7 @@ pub struct PyRoutingTarget(pub Option<RoutingTarget>);
 #[pymethods]
 impl PyRoutingTarget {
     #[staticmethod]
-    fn from_target(target: &Target) -> PyResult<Self> {
+    pub(crate) fn from_target(target: &Target) -> PyResult<Self> {
```
We should probably add the `from_target` method to `RoutingTarget` and have that be properly `pub` and `pub use`'d, rather than exposing `PyRoutingTarget` to the rest of Rust.

*edit*: we can follow up on this - we're going to need to rework the Sabre interfaces to use `TranspileLayout` anyway.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+                    )
+                })
+                .collect();
+            transpile_layout.compose_output_permutation(&routing_permutation, true);
```
This is quite likely already buggy - it doesn't handle relabelling of the existing output permutation that might have been induced by `ElidePermutations` on setting the `initial_layout`.

jakelishman replied:
Fixed in bea12c46e3132485269098e142f5bbfc68a63b3e.

jakelishman commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -35,12 +35,12 @@ use qiskit_circuit::imports::TRANSPILE_LAYOUT;
     /// transpilation. This vector maps the qubits at the start of the circuit to their
     /// final position/physical qubit at the end of the circuit.
-    output_permutation: Option<Vec<Qubit>>,
+    pub output_permutation: Option<Vec<Qubit>>,
```
Making this `pub` is just asking for us to make mistakes in handling the relabelling.  If you need to set the `initial_layout` manually, let's add a variant of `relabel_initial_layout` or something that takes in a new `NLayout` (or `fn(VirtualQubit) -> PhysicalQUbit` or whatever) and makes it impossible to forget to update/account for `output_permutation`.

jakelishman replied:
I can do this, btw

mtreinish replied:
The piece we don't have a story for yet is sabre where it internally applies an initial layout but doesn't update the `TranspileLayout` object . This is one difference from the model you added to python because we don't have the property set collection of layout components here, just the `TranspileLayout`. I agree we don't want this pub but this points out a gap in our `TranspilsLayout` api. I'll have to add a method to set the initial layout and update the existing permutation if it exists.

jakelishman replied:
Fixed in bea12c46e3132485269098e142f5bbfc68a63b3e.

jakelishman commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+            options.optimization_level
+        );
+    }
+    let seed = if options.seed < 0 {
+        None
+    } else {
+        Some(options.seed as u64)
+    };
```
minor: this is also `seed.try_into().ok()`.

jakelishman commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+///   execution (return code 0) the output of the transpiler will be written to the pointer
+/// @param error A pointer to a pointer with an nul terminated string with an error description.
+///   If the transpiler fails a pointer to the string with the error description will be written
+///   to this pointer. That pointer needs to be freed with `qk_str_free`.
```
I could see an argument for allowing `error` to be null, in which case we simply don't allocate and write out the "detailed" error message.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14760/commits/3a29e4c0a151a3934678ca775d9df67b3848d92e

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+) -> Result<(CircuitData, TranspileLayout)> {
+    if !(0..=3u8).contains(&optimization_level) {
+        panic!("Invalid optimization level specified {optimization_level}");
+    }
```
this might be an indication that an `enum` would be a better Rust-facing type

mtreinish replied:
I waffled back and forth on this I ended up on this because I liked how a u8 looked in c better, mainly I didn't like `QkTranspilerOptLevel_2 = 2` being exposed to the header file. I know it didn't have to be used or exposed to c even if we use an enum for rust but I felt it was easier to keep the interfaces the same. I can change it to internally use an enum if you'd prefer though.

jakelishman replied:
I agree with `u8` in C, just saying that the Rust side of it in `qiskit-transpiler` might want to be an enum.  But I also don't mind strongly.

mtreinish replied:
I switched the rust interface to use an enum in: https://github.com/Qiskit/qiskit/pull/14760/commits/85597f32518433aacea577dd8809fe7fea4e262c

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    );
+
+    let unroll_3q_or_more = |dag: &mut DAGCircuit| -> Result<()> {
+        // This will panic if there is a 3q unitary until qsd is ported
+        let mut out_dag = run_unitary_synthesis(
+            dag,
+            (0..dag.num_qubits()).collect(),
+            3,
+            Some(target),
+            HashSet::new(),
+            ["unitary".to_string(), "swap".to_string()]
+                .into_iter()
+                .collect(),
+            HashSet::new(),
+            approximation_degree,
+            None,
+            None,
+            false,
+        )?;
```
Panic?  We should probably return an error rather than abort the process.

mtreinish replied:
The whole error handling here needs a rework for everything. It was either returning a `PyErr` where we lose the error message and only return a backtrace to c or panic with an error message. I was also hoping https://github.com/Qiskit/qiskit/pull/14797 would be ready before we got to this point.

ShellyGarion replied:
#14797 (QSD in rust) is ready for review now, so hopefully could be merged for 2.2

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    }
+    // layout stage
+
+    let sabre_heuristic = sabre::Heuristic::new(
+        None,
+        None,
+        None,
+        target.num_qubits.map(|x| (x * 10) as usize),
+        1e-10,
+    )
+    .with_basic(1.0, sabre::SetScaling::Constant)
+    .with_lookahead(0.5, 20, sabre::SetScaling::Size)
+    .with_decay(0.001, 5)?;
```
Ha, I guess I never actually made a nice Rust-space interface.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+            |x| PhysicalQubit(x.0),
+        );
+    } else if optimization_level == 1 {
+        if run_check_map(&dag, target).is_none() {
+            apply_layout(
+                &mut dag,
+                &mut transpile_layout,
+                target.num_qubits.unwrap(),
+                |x| PhysicalQubit(x.0),
+            );
```
This typing of `run_check_map` is slightly weird - I think logically it actually is a `Result<(), E>`?

mtreinish replied:
Yeah I guess the rust interface is a bit too focused on the python interface usage where an option makes sense. I'll throw a comment here (and on the other usage) because I've already gotten this backwards once.

mtreinish replied:
Comment added in: https://github.com/Qiskit/qiskit/pull/14760/commits/35915b6860608dfabb41e5eba22d4d401824b843

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+                    )
+                })
+                .collect();
+            transpile_layout.compose_output_permutation(&routing_permutation, true);
```
Same trouble here with respect to `ElidePermutations`.  Let's not set fields on `TranspileLayout` directly.

jakelishman replied:
Fixed in bea12c46e3132485269098e142f5bbfc68a63b3e.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+            })
+            .collect();
+
+        transpile_layout.compose_output_permutation(&routing_permutation, true);
```
etc.  Also, let's factor this out into `TranspileLayout::compose_output_permutation_with_final_layout` or something like that.

jakelishman replied:
Fixed in bea12c46e3132485269098e142f5bbfc68a63b3e.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+        transpile_layout.compose_output_permutation(&routing_permutation, true);
+    }
+    // Routing stage
+    //    let vf2_post_result =
```
```suggestion
```

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    //        update_layout(&mut dag, &mut transpile_layout, |qubit| {
+    //            Qubit(post_layout[VirtualQubit(qubit.0)].0)
+    //        });
+    //    }
```
```suggestion
    }
```

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    //            Qubit(post_layout[VirtualQubit(qubit.0)].0)
+    //        });
+    //    }
+    // Translation Stage
+    let translation = |dag: &mut DAGCircuit, equiv_lib: &mut EquivalenceLibrary| -> Result<()> {
```
tbh, as a whole, the function could have been split into (internal) separate functions for each of the stages, just to let us more easily see the splits and the edges of the concerns.

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+                    (new_depth, new_size) != (self.best_depth, self.best_size)
+                }
+            }
+        }
```
This might be nicer separated out of this (very long!) function and with some comments on the functions.  It wasn't clear to me what `check` is returning until I see the name of the variable it's assigning to.

Do `new_size` and `new_depth` have to allow `Option`?  Seems like we've got two separate implementations of a do-while loop going here: one in `Option<usize>` for depth/size, and one in the `continue_loop` variable for this struct.  Feels like this could be a bit simpler, perhaps.

*edit*: potentially even, can it just be `MinState::update_with(&DAGCircuit) -> bool` or something like that?

mtreinish replied:
Long, it's less than 500 lines, if I move the struct out it'll be less than 400 :p . But yeah it makes more sense outside the function definition. Originally this was a closure and it needed to be here to capture the local variables on each call, but that was a nightmare to work with so I made this struct instead but just left it where the closure was.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14760/commits/3bbd9d39f0c82eb9b96ce5e03cc784f0d59307ee

jakelishman commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    Ok((dag_to_circuit(&dag, false)?, transpile_layout))
+}
+
+#[cfg(all(test, not(miri)))]
```
Why not Miri?  Can we comment - is it speed or can Miri not track the pointers / what?

mtreinish replied:
It's tracking pointers in `PackedOperation`, we can't use miri on any test that uses a `PackedOperation` right now. I actually wrote the rust space tests originally with the hope of having miri coverage but had to add this because I wasn't able to get it working.

jakelishman replied:
After the release, I'll rewrite `PackedOperation` in a way that should make Miri happy again - it's quite possible, especially now we only support 64-bit platforms.

jakelishman commented on docs/cdoc/qk-transpiler.rst:
```diff
@@ -0,0 +1,25 @@
+work solely in the case of a standalone C API. It will potentially not work
+correctly when in a mixed Python/C use case. If you're mixing C and Python you
+should call the :py:func:`.generate_preset_pass_manager` or
+:py:func:`.transpile` functions for those circuits.
```
```suggestion
The :c:func:`qk_transpiler` function exposes Qiskit's tranpsiler (:py:mod:`qiskit.transpiler`) to C.
The basic functionality is using the same underlying code as the Python-space version, but the transpiler as
exposed to C has more limitations than what is exposed to Python. The transpiler
assumes a circuit built constructed using solely the C API and is intended to
work solely in the case of a standalone C API. It will potentially not work
correctly when in a mixed Python/C use case. If you're mixing C and Python you
should call the :py:func:`.generate_preset_pass_manager` or
:py:func:`.transpile` functions for those circuits.
```

Take whichever suggestions you want from this - at the moment, we're doing a weird mix of explicit Python-domain cross references but not using C-domain cross references, and some of the sentences needed a bit of a tweak.

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+    qk_target_add_instruction(target, ecr_entry);
+
+    QkCircuit *qc = qk_circuit_new(num_qubits, 0);
+    uint32_t x_qargs[1] = {
+        9,
+    };
```
Is it the trailing commas that made clang-format split these lines?  Since it didn't for the 2q ones.

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+    };
+    qk_circuit_gate(qc, QkGate_X, x_qargs, NULL);
+    for (uint32_t i = 0; i < qk_circuit_num_qubits(qc); i++) {
+        uint32_t qargs[1] = {
+            i,
+        };
+        qk_circuit_gate(qc, QkGate_H, qargs, NULL);
```
very minor: `qk_circuit_gate(qc, QkGate_H, &i, NULL);` would work too for 1q gates.

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+    for (uint32_t i = 0; i < qk_circuit_num_qubits(qc) - 1; i += 2) {
+        uint32_t qargs[2] = {i, 9};
+        qk_circuit_gate(qc, QkGate_CX, qargs, NULL);
+    }
```
This is a weird mix of dynamic circuit length (in the loop) and a hard-coded number 9.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14760/commits/16b6a98645109350a7c0267da195bbb2bb06d0ba

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+        printf("Transpilation failed with: %s\n", error);
+        result = EqualityError;
+        goto circuit_cleanup;
+    }
```
We need to free `error` in this path.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14760/commits/16b6a98645109350a7c0267da195bbb2bb06d0ba

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+            printf("Transpilation failed %s\n", error);
+            result = EqualityError;
+            goto cleanup;
+        }
```
Need to free `error`.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14760/commits/16b6a98645109350a7c0267da195bbb2bb06d0ba

jakelishman commented on test/c/test_transpiler.c:
```diff
@@ -0,0 +1,209 @@
+    fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);
+
+    return num_failed;
+}
```
Seems like we're not actually running the idle qubits test?

mtreinish replied:
Oops I was very surprised the gate count was the same for each optimization level, that would probably explain it lol

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14760/commits/5c96d44c3533aacb5d89f0a1ca4bfc0b528e5697

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+    /// The optimization level to run the transpiler with. Valid values are 0, 1, 2, or 3.
+    optimization_level: u8,
+    /// The seed for the transpiler. If set to a negative number this means no seed will be
+    /// set and the RNGs used in the transpiler will be seeded froms system entropy.
```
```suggestion
    /// set and the RNGs used in the transpiler will be seeded from system entropy.
```

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+}
+
+/// @ingroup QkTranspiler
+/// Transpile a single circuit that was constructed using the C API
```
Since the C API ist an equally valid API as the Python one it sounds a bit strange to me to explicitly say C API here. What do you think about just saying `Transpile a single circuit` and note below somewhere that currently the circuit can only have operations supported in the C API and giving a circuit constructed from a Python API can break?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14760/commits/e5cbe55511ff1f669401b482411b22d33706df5f

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+/// @param circuit A pointer to the circuit to run the transpiler on
+/// @param target A pointer to the target to compile the circuit for
+/// @params options A pointer to an options object that define user options if this is a null
+///   pointer the default values will be used
```
```suggestion
/// @params options A pointer to an options object that defines user options. If this is a null
///   pointer the default values will be used. See ``qk_transpile_default_options``
///   for more details on the default values.
```

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+    options: *const TranspileOptions,
+    result: *mut TranspileResult,
+    error: *mut *mut c_char,
+) -> i32 {
```
Could we use `ExitCode` here? E.g. the `ExitCode::TranspilerError` could be the generic error to return upon failure

mtreinish replied:
Sure, I'll add it to the enum and add a `TranspilerError` variant.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14760/commits/e5cbe55511ff1f669401b482411b22d33706df5f

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+    target: *const Target,
+    options: *const TranspileOptions,
+    result: *mut TranspileResult,
+    error: *mut *mut c_char,
```
What do you think about calling this `backtrace` to be explicit about what it is? A variable called `error` could intuitively also only mean the error code as integer -- just a suggestion

jakelishman replied:
The fact that it's currently a backtrace is just an implementation detail of what was possible to get without accidentally requiring the GIL - it's only supposed to be a message.

Cryoris commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,181 @@
+///
+/// Behavior is undefined if ``circuit``, ``target``, ``options``, ``result``, or ``error`` are
+/// not valid, non-null pointers to a ``QkCircuit``, ``QkTarget``, ``QkTranspileOptions``,
+/// ``QkTranspileResult``, or a ``char`` pointer respectively.
```
It should be a character pointer pointer, right?
```suggestion
/// ``QkTranspileResult``, or a ``char**`` respectively.
```

jakelishman replied:
The PR text is correct - `error` must be a (valid, non-null pointer) to a (`char` pointer).

(Though arguably, we could easily handle the case that it _is_ a null pointer and just not write out the error message.)

Cryoris commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+        panic!("Invalid optimization level specified {optimization_level}");
+    }
+    let mut dag = DAGCircuit::from_circuit_data(circuit, false, None, None, None, None)?;
+    let mut commutation_checker = get_standard_commutation_checker();
+    let mut equivalence_library = generate_standard_equivalence_library();
```
Should these be cached in case we call `transpile` repeatedly on a large number of circuits?

mtreinish replied:
There really isn't a good story for caching these right now. My thinking longer term is that we have a lot of one time setup in the transpiler (passes building views of the target etc) that we will want to expose a `PassManager` struct to C that is used for persisting the state between multiple runs. With a single call function like this there really isn't a good pattern for persisting this right now. The bigger issue is that the equivalence library needs to be mutable because of how the search works (a dummy node is added to the graph to serve as a source point for the path search). The commutation library we could potentially use a global for, but it is mostly static and ends up basically being a big hashmap allocation and should be sufficiently fast to not warrant the complexity.

For the multiple circuits case, one idea I had which wouldn't be hard to do is a `transpile_circuits()` function which takes in a `&[CircuitData]` and basically does `circuits.par_iter().map(|x| transpile(x, ...)).collect()`, we could potentially reuse things that way but would need to put the equivalence library behind a mutex or something so whenever we call the basis translator only one circuit is running at a time. I didn't want to include that in this PR though since it is large enough already.

jakelishman replied:
> The bigger issue is that the equivalence library needs to be mutable because of how the search works (a dummy node is added to the graph to serve as a source point for the path search).

We should be able to fix this, especially now it's all in Rust space - if nothing else, we can just hard-code the first "step" of the search ourselves, if not modify Rustworkx's interface slightly to more easily let us do what we want.

Cryoris commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+    if optimization_level == 1 {
+        run_inverse_cancellation_standard_gates(&mut dag);
+    } else if optimization_level == 2 || optimization_level == 3 {
+        if let Some((new_dag, permutation)) = run_elide_permutations(&dag)? {
```
It's a bit weird that not all passes run run in-place consistently (this one and split_2q_unitaries), we should make this consistent (not here ofc)

mtreinish replied:
The passes that return a new dag typically need to rewrite the whole dag which is more efficient to do from scratch than mutating in place. I guess they could do what we do in the C API and write the new dag into a mutable reference but I'm not sure how much it matters. I kind of like knowing that the output is explicitly a copy and could potentially reuse the original dag.

That being said for `split_2q_unitaries` it can actually be either, it will mutate in place if there are only unitaries that are the tensor product of single qubit unitaries, While it will return a copy if any unitary is swap equivalent because it needs to rewrite the circuit to elide the swap.

mtreinish committed:
Remove commented out vf2postlayout code

Co-authored-by: Jake Lishman <jake@binhbar.com>

Cryoris commented on crates/transpiler/src/transpiler.rs:
```diff
@@ -0,0 +1,710 @@
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::{PhysicalQubit, Qubit};
+
+/// A transpilation function for Rust native circuits for use in the C API. This will not cover
```
It would be nice to document on the C API which passes are run, or at least the more important ones like what routing algo is used. But we can do that in a follow up too

mtreinish replied:
It does match the default Python behavior with the exception of VF2PostLayout which isn't in rust yet or passes that don't apply to C (like HLS). I can write this out, but I was hoping to mostly refer people to the transpiler module docs which explain things in more detail.

Cryoris commented on docs/cdoc/qk-transpiler.rst:
```diff
@@ -0,0 +1,25 @@
+Data Types
+==========
+
+.. doxygenstruct:: QkTranspileOptions
```
This seems to be missing `QkTranspileResult`

mtreinish replied:
I had it there originally but I got a duplicate struct member error when I included it here. I'll play with it again locally to see what's going on.

mtreinish replied:
I figured it out, the Result struct was listed as in the group already so the functions list below was picking it up. So if I manually added it here it was showing up as a duplicate. Removing the group include from the docstring fixes that.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14760/commits/f0251fbbe556442eb10db6fa7da9bf6080a0713d

Cryoris reviewed (commented):
Very nice! Some initial (mostly doc) comments 🙂

mtreinish committed:
Run idle qubits test

mtreinish committed:
Add release note

mtreinish committed:
Move MinPointState definition outside transpile func and simplify interface

mtreinish committed:
Merge branch 'main' into transpile-c-wip

mtreinish committed:
Docs updates

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>
Co-authored-by: Jake Lishman <jake@binhbar.com>

jakelishman committed:
Add permutation-mutation methods to `TranspileLayout`

This rewrites how permutations are mutated within the `TranspileLayout`
structure, avoiding exposing the internal state (which is desperately
easy to update incorrectly when done piecewise) and instead exposing
methods that let us either add in a new permutation between the circuit
and the existing tracked layout/permutation (the typical thing we're
doing), or to include (only) a permutation that was being tracked before
this new `TranspileLayout` object was created.

The `add_permutation_outside` method may well be obsoleted in the future
once Sabre is properly acting on `TranspileLayout` itself.

mtreinish committed:
Allow error to be null and just don't write out error msg

mtreinish committed:
Use an enum for optimization level

mtreinish committed:
Fix c test issues

mtreinish committed:
Use ExitCode enum for C return type

mtreinish committed:
Add comment about run_check_map return

mtreinish committed:
Fix rust tests

raynelfss commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,200 @@
+///   members of the result struct are owned by the caller and you are responsible for freeing
+///   the members using the respective free functions.
+/// @param error A pointer to a pointer with an nul terminated string with an error description.
+///   If the transpiler fails a pointer to the string with the error description will be written
```
```suggestion
///   If the transpiler fails, a pointer to the string with the error description will be written
```
Also, should we adopt this pattern with the standalone passes in C?

mtreinish replied:
I think after we resolve the error handling story for C. Right now this pattern is kind of future facing and we just stuff a backtrace in the error string as that's the best we can do.

raynelfss commented on crates/cext/src/transpiler/transpile_function.rs:
```diff
@@ -0,0 +1,200 @@
+///
+/// @param circuit A pointer to the circuit to run the transpiler on
+/// @param target A pointer to the target to compile the circuit for
+/// @params options A pointer to an options object that defines user options. If this is a null
```
```suggestion
/// @params options A pointer to a ``QkTranspileOptions`` object that defines user options. If this is a null
```

raynelfss reviewed (commented):
Minor comments/questions from quick pass through the code.

jakelishman committed:
Fix typing of tests

mtreinish committed:
Add QkTranspileResult to the correct docs section

mtreinish committed:
Fix docs typo in transpiler page preamble

mtreinish committed:
Merge branch 'main' into transpile-c-wip

mtreinish committed:
Fix merge conflict with function rename in tests

jakelishman reviewed (approved):
I'm happy enough to merge this for 2.2.0b1 and to continue further work in separated PRs.
